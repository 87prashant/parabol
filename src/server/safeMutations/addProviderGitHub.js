import {toGlobalId} from 'graphql-relay';
import fetch from 'node-fetch';
import {stringify} from 'querystring';
import getRethink from 'server/database/rethinkDriver';
import tokenCanAccessRepo from 'server/integrations/tokenCanAccessRepo';
import getProviderRowData from 'server/safeQueries/getProviderRowData';
import postOptions from 'server/utils/fetchOptions';
import getPubSub from 'server/utils/getPubSub';
import makeAppLink from 'server/utils/makeAppLink';
import shortid from 'shortid';
import {GITHUB, GITHUB_ENDPOINT, GITHUB_SCOPE} from 'universal/utils/constants';
import makeGitHubPostOptions from 'universal/utils/makeGitHubPostOptions';
import maybeJoinRepos from 'server/integrations/githubWebhookHandlers/maybeJoinRepos';

const profileQuery = `
query { 
  viewer { 
    login
    id
    organizations(first: 100) {
      nodes {
        login
        viewerCanAdminister
      }
    }
  }
}`;

const createOrgWebhooks = async (accessToken, organizations) => {
  const orgsWithAdminRights = organizations.nodes.filter((org) => org.viewerCanAdminister);
  // for each org, see if the webhook already exists
  // if it does, go on to next
  // if it does not, create it

  const webhookLists = await Promise.all(orgsWithAdminRights.map((org) => {
    const endpoint = `https://api.github.com/orgs/${org.login}/hooks`;
    return fetch(endpoint, {headers: {Authorization: `Bearer ${accessToken}`}}).then((res) => res.json());
  }));
  const createHookParams = {
    name: 'web',
    config: {
      url: makeAppLink('webhooks/github', {isWebhook: true}),
      content_type: 'json',
      // this doesn't have to be the client secret, but i don't see much harm in reusing it, assuming it's all SSL
      secret: process.env.GITHUB_CLIENT_SECRET,
      insecure_ssl: 0
    },
    events: ['organization'],
    active: true
  };

  webhookLists.forEach((hookList, idx) => {
    if (hookList.length === 0) {
      const org = orgsWithAdminRights[idx];
      const endpoint = `https://api.github.com/orgs/${org.login}/hooks`;
      fetch(endpoint, makeGitHubPostOptions(accessToken, createHookParams))
        .then((body) => body.json())
        .then((res) => console.log('res', res))
    }
  });

};

const getJoinedIntegrationIds = async (integrationCount, accessToken, teamId, userId) => {
  if (integrationCount === 0) return [];
  const r = getRethink();
  const allIntegrations = await r.table(GITHUB)
    .getAll(teamId, {index: 'teamId'})
    .filter({isActive: true});
  return maybeJoinRepos(allIntegrations, accessToken, userId);
};

const getTeamMember = async (joinedIntegrationIds, teamMemberId) => {
  if (joinedIntegrationIds.length > 0) {
    const r = getRethink();
    return r.table('TeamMember').get(teamMemberId).pluck('id', 'preferredName', 'picture');
  }
  return undefined;
};

const addProviderGitHub = async (code, teamId, userId) => {
  const r = getRethink();
  const now = new Date();
  const queryParams = {
    client_id: process.env.GITHUB_CLIENT_ID,
    client_secret: process.env.GITHUB_CLIENT_SECRET,
    code
    // redirect_uri: makeAppLink('auth/github/entry')
  };
  const uri = `https://github.com/login/oauth/access_token?${stringify(queryParams)}`;
  const ghRes = await fetch(uri, postOptions);
  const json = await ghRes.json();
  const {access_token: accessToken, error, scope} = json;
  if (error) {
    throw new Error(`GitHub: ${error}`);
  }
  if (scope !== GITHUB_SCOPE) {
    throw new Error(`bad scope: ${scope}`);
  }
  const authedPostOptions = makeGitHubPostOptions(accessToken, {
    query: profileQuery
  })
  const ghProfile = await fetch(GITHUB_ENDPOINT, authedPostOptions);
  const gqlRes = await ghProfile.json();
  if (!gqlRes.data) {
    console.error('GitHub error: ', gqlRes);
  }
  const {data: {viewer: {login, organizations}}} = gqlRes;
  const provider = await r.table('Provider')
    .getAll(teamId, {index: 'teamIds'})
    .filter({service: GITHUB, userId})
    .nth(0)('id')
    .default(null)
    .do((providerId) => {
      return r.branch(
        providerId.eq(null),
        r.table('Provider')
          .insert({
            id: shortid.generate(),
            accessToken,
            createdAt: now,
            // github userId is never used for queries, but the login is!
            providerUserId: login,
            providerUserName: login,
            service: GITHUB,
            teamIds: [teamId],
            updatedAt: now,
            userId
          }, {returnChanges: true})('changes')(0)('new_val'),
        r.table('Provider')
          .get(providerId)
          .update({
            accessToken,
            updatedAt: now,
            providerUserId: login,
            providerUserName: login
          }, {returnChanges: true})('changes')(0)('new_val')
      );
    });
  createOrgWebhooks(accessToken, organizations);

  const rowDetails = await getProviderRowData(GITHUB, teamId);
  const joinedIntegrationIds = await getJoinedIntegrationIds(rowDetails.integrationCount, accessToken, teamId, userId)
  const teamMemberId = `${userId}::${teamId}`;
  const teamMember = await getTeamMember(joinedIntegrationIds, teamMemberId);
  const providerAdded = {
    provider,
    providerRow: {
      ...rowDetails,
      accessToken,
      service: GITHUB,
      // tell relay to not automatically merge the new value as a sink. changed teamId changes globalId
      teamId: `_${teamId}`
    },
    joinedIntegrationIds,
    teamMember
  };

  getPubSub().publish(`providerAdded.${teamId}`, {providerAdded});
};

export default addProviderGitHub;
