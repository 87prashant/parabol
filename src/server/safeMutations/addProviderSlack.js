import fetch from 'node-fetch';
import {stringify} from 'querystring';
import getRethink from 'server/database/rethinkDriver';
import postOptions from 'server/utils/fetchOptions';
import getPubSub from 'server/utils/getPubSub';
import makeAppLink from 'server/utils/makeAppLink';
import shortid from 'shortid';
import {SLACK, SLACK_SCOPE} from 'universal/utils/constants';
import getProviderRowData from 'server/safeQueries/getProviderRowData';

const addProviderSlack = async (code, teamId, userId) => {
  const r = getRethink();
  const now = new Date();
  const queryParams = {
    client_id: process.env.SLACK_CLIENT_ID,
    client_secret: process.env.SLACK_CLIENT_SECRET,
    code,
    redirect_uri: makeAppLink('auth/slack')
  };
  const uri = `https://slack.com/api/oauth.access?${stringify(queryParams)}`;
  const slackRes = await fetch(uri, postOptions);
  const json = await slackRes.json();
  const {ok, error, access_token: accessToken, scope, team_id: providerUserId, team_name: providerUserName} = json;
  if (!ok) {
    throw new Error(error);
  }
  if (scope !== SLACK_SCOPE) {
    throw new Error(`bad scope: ${scope}`);
  }
  const provider = await r.table('Provider')
    .getAll(teamId, {index: 'teamIds'})
    .filter({service: SLACK})
    .nth(0)('id')
    .default(null)
    .do((providerId) => {
      return r.branch(
        providerId.eq(null),
        r.table('Provider')
          .insert({
            id: shortid.generate(),
            accessToken,
            createdAt: now,
            providerUserId,
            providerUserName,
            service: SLACK,
            teamIds: [teamId],
            updatedAt: now,
            userId
          }, {returnChanges: true})('changes')(0)('new_val'),
        r.table('Provider')
          .get(providerId)
          .update({
            accessToken,
            updatedAt: now
          }, {returnChanges: true})('changes')(0)('new_val')
      );
    });

  // need to fetch this because it could be a refresh
  const rowDetails = await getProviderRowData(SLACK, teamId);

  const providerAdded = {
    provider,
    providerRow: {
      ...rowDetails,
      accessToken,
      service: SLACK,
      teamId
    }
  };
  getPubSub().publish(`providerAdded.${teamId}`, {providerAdded});
};

export default addProviderSlack;