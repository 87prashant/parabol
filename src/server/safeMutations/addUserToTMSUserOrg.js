import getRethink from 'server/database/rethinkDriver';

const addUserToTMSUserOrg = (userId, teamId, orgId, options = {}) => {
  const {returnChanges, role = null} = options;
  const r = getRethink();
  return r.table('User')
    .get(userId)
    .update((userDoc) => ({
      userOrgs: r.branch(
        userDoc('userOrgs').contains((userOrg) => userOrg('id').eq(orgId)).default(false),
        userDoc('userOrgs'),
        userDoc('userOrgs').append({
          id: orgId,
          role
        })
      ),
      tms: r.branch(
        userDoc('tms').contains(teamId).default(false),
        userDoc('tms'),
        userDoc('tms').default([]).append(teamId)
      )
    }), {returnChanges});
};

export default addUserToTMSUserOrg;
