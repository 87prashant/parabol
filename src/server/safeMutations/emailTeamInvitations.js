import getRethink from 'server/database/rethinkDriver';
import createEmailPromises from '../graphql/models/Invitation/inviteTeamMembers/createEmailPromises';
import makeInvitationsForDB from '../graphql/models/Invitation/inviteTeamMembers/makeInvitationsForDB';
import makeInviteToken from '../graphql/models/Invitation/inviteTeamMembers/makeInviteToken';
import resolveSentEmails from '../graphql/models/Invitation/inviteTeamMembers/resolveSentEmails';

export default async function emailTeamInvitations(invitees, inviter) {
  if (invitees.length === 0) return invitees;
  const {teamId, userId: inviterUserId} = inviter;
  const r = getRethink();
  const inviteesWithTokens = invitees.map((invitee) => ({...invitee, inviteToken: makeInviteToken()}));
  const sendEmailPromises = createEmailPromises(inviter, inviteesWithTokens);
  const {inviteesToStore} = await resolveSentEmails(sendEmailPromises, inviteesWithTokens);
  const invitationsForDB = await makeInvitationsForDB(inviteesToStore, teamId, inviterUserId);
  // Bulk insert, wait in case something queries the invitation table
  return r.table('Invitation').insert(invitationsForDB);
  // TODO generate email to inviter including folks that we couldn't reach
  // if (inviteeErrors.length > 0) {
  // throw errorObj({_error: 'Some invitations were not sent', type: 'inviteSendFail', failedEmails: inviteeErrors});
  // }
};
