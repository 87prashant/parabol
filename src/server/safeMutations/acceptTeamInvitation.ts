import adjustUserCount from 'server/billing/helpers/adjustUserCount'
import getRethink from 'server/database/rethinkDriver'
import addTeamMemberToNewMeeting from 'server/graphql/mutations/helpers/addTeamMemberToNewMeeting'
import addUserToTMSUserOrg from 'server/safeMutations/addUserToTMSUserOrg'
import insertNewTeamMember from 'server/safeMutations/insertNewTeamMember'
import {ADD_USER} from 'server/utils/serverConstants'
import {TEAM_INVITATION} from 'universal/utils/constants'
import {auth0ManagementClient} from 'server/utils/auth0Helpers'

const acceptTeamInvitation = async (
  teamId: string,
  userId: string,
  invitationId: string,
  dataLoader: any
) => {
  const r = getRethink()
  const now = new Date()
  const {
    team: {orgId},
    user
  } = await r({
    team: r.table('Team').get(teamId),
    user: r.table('User').get(userId)
  })
  const userOrgs = user.userOrgs || []
  const userInOrg = Boolean(userOrgs.find((org) => org.id === orgId))
  const {removedNotificationIds, teamMember} = await r({
    // add the team to the user doc
    userUpdate: addUserToTMSUserOrg(userId, teamId, orgId),
    teamMember: insertNewTeamMember(userId, teamId),
    // only redeem 1 invitation. any others will expire
    redeemedInvitation: r
      .table('TeamInvitation')
      .get(invitationId)
      .update({
        acceptedAt: now,
        acceptedBy: userId
      }),
    removedNotificationIds: r
      .table('Notification')
      .getAll(userId, {index: 'userIds'})
      .filter({
        type: TEAM_INVITATION,
        teamId
      })
      .update({isArchived: true}, {returnChanges: true})('changes')('new_val')('id')
  })
  if (!userInOrg) {
    await adjustUserCount(userId, orgId, ADD_USER)
  }

  // if a meeting is going on right now, add them
  await addTeamMemberToNewMeeting(teamMember, teamId, dataLoader)

  // update auth0
  const tms = user.tms ? user.tms.concat(teamId) : [teamId]
  auth0ManagementClient.users.updateAppMetadata({id: userId}, {tms})
  return removedNotificationIds
}

export default acceptTeamInvitation
