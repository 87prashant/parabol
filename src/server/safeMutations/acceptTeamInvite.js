import adjustUserCount from 'server/billing/helpers/adjustUserCount';
import getRethink from 'server/database/rethinkDriver';
import addUserToTMSUserOrg from 'server/safeMutations/addUserToTMSUserOrg';
import insertNewTeamMember from 'server/safeMutations/insertNewTeamMember';
import getTeamInviteNotifications from 'server/safeQueries/getTeamInviteNotifications';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import {getUserId} from 'server/utils/authorization';
import getPubSub from 'server/utils/getPubSub';
import {ADD_USER} from 'server/utils/serverConstants';
import tmsSignToken from 'server/utils/tmsSignToken';
import {
  ADD_TO_TEAM, ADDED, INVITATION_REMOVED, JOIN_TEAM, NEW_AUTH_TOKEN, NOTIFICATIONS_ADDED, NOTIFICATIONS_CLEARED,
  TEAM_ADDED, TEAM_MEMBER
} from 'universal/utils/constants';
import toTeamMemberId from 'universal/utils/relay/toTeamMemberId';

const publishRemovedInvitations = (expiredEmailInvitations, operationId) => {
  expiredEmailInvitations.forEach((invitation) => {
    const {teamId} = invitation;
    const invitationRemoved = {invitation};
    getPubSub().publish(`${INVITATION_REMOVED}.${teamId}`, {invitationRemoved, operationId});
  });
};

const publishClearNotifications = (expireInviteNotificationIds, userId) => {
  if (expireInviteNotificationIds.length > 0) {
    const notificationsCleared = {deletedIds: expireInviteNotificationIds};
    getPubSub().publish(`${NOTIFICATIONS_CLEARED}.${userId}`, {notificationsCleared});
  }
};

const publishJoinTeamNotifications = (teamId, teamName, user) => {
  const notificationsAdded = {
    notifications: [{
      type: JOIN_TEAM,
      teamId,
      teamName,
      preferredName: user.preferredName || user.email
    }]
  };
  getPubSub().publish(`${NOTIFICATIONS_ADDED}.${teamId}`, {notificationsAdded});
};

const acceptTeamInvite = async (teamId, authToken, email, subOptions = {}) => {
  const r = getRethink();
  const now = new Date();
  const {operationId} = subOptions;
  const userId = getUserId(authToken);
  const teamMemberId = toTeamMemberId(teamId, userId);
  const {team: {orgId, name: teamName}, user} = await r({
    team: r.table('Team').get(teamId).pluck('orgId', 'name'),
    user: r.table('User').get(userId)
  });
  const userOrgs = user.userOrgs || [];
  const userTeams = user.tms || [];
  const userInOrg = Boolean(userOrgs.find((org) => org.id === orgId));
  const tms = [...userTeams, teamId];
  const {expiredEmailInvitations, expireInviteNotificationIds, teamMember} = await r({
    // add the team to the user doc
    userUpdate: addUserToTMSUserOrg(userId, teamId, orgId),
    teamMember: insertNewTeamMember(userId, teamId),
    // find all possible emails linked to this person and mark them as accepted
    expiredEmailInvitations: r.table('Invitation')
      .getAll(email, {index: 'email'})
      .update({
        acceptedAt: now,
        // flag the token as expired so they cannot reuse the token
        tokenExpiration: new Date(0),
        updatedAt: now
      }, {returnChanges: true})('changes')('new_val').default([]),
    expireInviteNotificationIds: getTeamInviteNotifications(orgId, teamId, [email])
      .delete({returnChanges: true})('changes')('old_val')('id')
      .default([])
  });

  if (!userInOrg) {
    await adjustUserCount(userId, orgId, ADD_USER);
  }
  auth0ManagementClient.users.updateAppMetadata({id: userId}, {tms});

  publishRemovedInvitations(expiredEmailInvitations, operationId);
  publishClearNotifications(expireInviteNotificationIds, userId);
  publishJoinTeamNotifications(teamId, teamName, user);

  getPubSub().publish(`${TEAM_MEMBER}.${teamId}`, {teamMember: {teamMemberId, type: ADDED}, ...subOptions});

  // Send the new team member a welcome & a new token
  const newAuthToken = tmsSignToken(authToken, tms);
  const addedToTeam = {
    // no ID to rule out permanent notification. they already know, they are the ones who did it!
    startAt: now,
    type: ADD_TO_TEAM,
    teamId
  };
  const teamAdded = {
    notification: addedToTeam,
    teamId,
    teamMemberId
  };
  getPubSub().publish(`${TEAM_ADDED}.${userId}`, {teamAdded, ...subOptions});
  getPubSub().publish(`${NEW_AUTH_TOKEN}.${userId}`, {newAuthToken});

  return {
    authToken: newAuthToken,
    teamId,
    teamMemberId
  };
};

export default acceptTeamInvite;
