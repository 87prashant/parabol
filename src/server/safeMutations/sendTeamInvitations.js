import getRethink from 'server/database/rethinkDriver';
import emailTeamInvitations from 'server/safeMutations/emailTeamInvitations';
import shortid from 'shortid';
import {TEAM_INVITE} from 'universal/utils/constants';

const sendTeamInvitations = async (invitees, inviter) => {
  if (invitees.length === 0) return [];
  const r = getRethink();
  const now = new Date();
  const {orgId, inviterName, teamId, teamName} = inviter;
  const invitations = invitees
    .filter((invitee) => invitee.isUser)
    .map((invitee) => ({
      id: shortid.generate(),
      type: TEAM_INVITE,
      startAt: now,
      orgId,
      userIds: [invitee.userId],
      inviterName,
      teamId,
      teamName
    }));

  await Promise.all([
    r.table('Notification').insert(invitations),
    emailTeamInvitations(invitees, inviter)
  ]);

  const notificationsToAdd = {};
  invitations.forEach((notification) => {
    notificationsToAdd[notification.userIds[0]] = [notification];
  });
  return notificationsToAdd;
};

export default sendTeamInvitations;
