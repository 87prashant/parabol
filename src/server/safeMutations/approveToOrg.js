import getRethink from 'server/database/rethinkDriver'
import sendTeamInvitations from 'server/safeMutations/sendTeamInvitations'
import {APPROVED, PENDING} from 'server/utils/serverConstants'
import shortid from 'shortid'
import {INVITEE_APPROVED, REQUEST_NEW_USER} from 'universal/utils/constants'
import standardError from 'server/utils/standardError'

const approveToOrg = async (email, orgId, userId, dataLoader) => {
  const r = getRethink()
  const now = new Date()
  // get all notifications for this email to join this org
  const removedNotifications = await r
    .table('Notification')
    .getAll(userId, {index: 'userIds'})
    .filter({
      orgId,
      type: REQUEST_NEW_USER,
      inviteeEmail: email
    })
    .delete({returnChanges: true})('changes')('old_val')
    .default([])
  if (removedNotifications.length === 0) {
    return standardError(new Error('Notification not found'), {userId})
  }
  // RESOLUTION
  // send 1 team invite per notification
  const inviterUserIds = Array.from(
    removedNotifications.reduce((userIds, notification) => {
      return userIds.add(notification.inviterUserId)
    }, new Set())
  )

  const teamIds = removedNotifications.map(({teamId}) => teamId)
  const {inviterUsers, teams} = await r({
    inviterUsers: r
      .table('User')
      .getAll(r.args(inviterUserIds), {index: 'id'})
      .pluck('id', 'email', 'picture', 'preferredName')
      .coerceTo('array'),
    teams: r
      .table('Team')
      .getAll(r.args(teamIds), {index: 'id'})
      .pluck('id', 'name')
      .coerceTo('array')
  })
  const inviters = removedNotifications.map((notification) => {
    const inviterUser = inviterUsers.find(({id}) => id === notification.inviterUserId)
    const team = teams.find(({id}) => id === notification.teamId)
    return {
      inviterUserId: inviterUser.id,
      inviterEmail: inviterUser.email,
      inviterName: inviterUser.preferredName,
      teamName: team.name,
      orgId,
      teamId: team.id,
      userId: inviterUser.id
    }
  })
  const inviteeApprovedNotifications = inviters.map((inviter) => {
    const {inviterName, userId: inviterUserId, teamId, teamName} = inviter
    return {
      id: shortid.generate(),
      type: INVITEE_APPROVED,
      startAt: now,
      orgId,
      userIds: [inviterUserId],
      inviteeEmail: email,
      inviterName,
      teamId,
      teamName
    }
  })

  // tell the inviters that their friend was approved
  // send the invitee a series of team invites
  const {inviteeUser, removedOrgApprovals} = await r({
    insertInviteeApproved: r.table('Notification').insert(inviteeApprovedNotifications),
    removedOrgApprovals: r
      .table('OrgApproval')
      .getAll(email, {index: 'email'})
      .filter({orgId, status: PENDING, isActive: true})
      .update(
        {
          status: APPROVED,
          approvedBy: userId,
          updatedAt: now
        },
        {returnChanges: true}
      )('changes')('new_val')
      .default([]),
    inviteeUser: r
      .table('User')
      .getAll(email, {index: 'email'})
      .nth(0)
      .default(null)
  })

  const invitees = inviteeUser ? [{email, userId: inviteeUser.id}] : [{email}]

  const sentTeamInvitations = await Promise.all(
    inviters.map((inviter) => {
      return sendTeamInvitations(invitees, inviter, undefined, dataLoader)
    })
  )

  const newInvitations = sentTeamInvitations.reduce((arr, upserts) => {
    arr.push(...upserts.newInvitations)
    return arr
  }, [])

  const teamInviteNotifications = sentTeamInvitations.reduce((arr, upserts) => {
    arr.push(...upserts.teamInviteNotifications)
    return arr
  }, [])

  return {
    removedRequestNotifications: removedNotifications,
    removedOrgApprovals,
    newInvitations,
    inviteeApprovedNotifications,
    teamInviteNotifications
  }
}

export default approveToOrg
