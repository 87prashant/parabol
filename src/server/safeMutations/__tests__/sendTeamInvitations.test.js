import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import getRethink from 'server/database/rethinkDriver';
import sendTeamInvitations from 'server/safeMutations/sendTeamInvitations';
import shortid from 'shortid';
import * as emailTeamInvitations from 'server/safeMutations/emailTeamInvitations';
import {TEAM_INVITE} from 'universal/utils/constants';

MockDate.set(__now);
console.error = jest.fn();

describe('sendTeamInvitations', () => {
  test('returns an empty array if there are no people to invite', async () => {
    // TEST
    const res = await sendTeamInvitations([], {});

    // VERIFY
    expect(res).toEqual([]);
  });
  test('inserts invitation notifications, returns an object full of the notifications removed', async () => {
    // SETUP
    emailTeamInvitations.default = jest.fn();
    const r = getRethink();
    const mockDB = new MockDB();
    const dynamicSerializer = new DynamicSerializer();
    const invitee = {userId: `invitee|${shortid.generate()}`};
    await mockDB.init();
    const firstUser = mockDB.db.user[0];
    const inviter = {
      inviterName: firstUser.preferredName,
      teamName: mockDB.context.team.name,
      teamId: mockDB.context.team.id,
      orgId: mockDB.context.organization.id
    };

    // TEST
    const res = await sendTeamInvitations([invitee], inviter);

    // VERIFY
    expect(emailTeamInvitations.default).toHaveBeenCalledTimes(1);
    expect(res[invitee.userId].length).toEqual(1);
    const db = await fetchAndSerialize({
      notification: r.table('Notification').getAll(inviter.orgId, {index: 'orgId'}).orderBy('userIds')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });
  test('does not add a duplicate notification', async () => {
    // SETUP
    emailTeamInvitations.default = jest.fn();
    const r = getRethink();
    const mockDB = new MockDB();
    const dynamicSerializer = new DynamicSerializer();
    const invitee = {userId: `invitee|${shortid.generate()}`};
    await mockDB.init()
      .newNotification({userIds: [invitee.userId]}, {type: TEAM_INVITE});
    const firstUser = mockDB.db.user[0];
    const inviter = {
      inviterName: firstUser.preferredName,
      teamName: mockDB.context.team.name,
      teamId: mockDB.context.team.id,
      orgId: mockDB.context.organization.id
    };

    // TEST
    const res = await sendTeamInvitations([invitee], inviter);

    // VERIFY
    expect(emailTeamInvitations.default).toHaveBeenCalledTimes(1);
    expect(res[invitee.userId].length).toEqual(1);
    const db = await fetchAndSerialize({
      notification: r.table('Notification').getAll(inviter.orgId, {index: 'orgId'}).orderBy('userIds')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });
});
