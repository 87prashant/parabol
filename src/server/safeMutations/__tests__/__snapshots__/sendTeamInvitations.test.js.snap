// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendTeamInvitations inserts invitation notifications, returns an object full of the notifications removed 1`] = `
Object {
  "notification": Array [
    Object {
      "id": 0,
      "inviterName": "Captain America",
      "orgId": 1,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamId": 2,
      "teamName": "Team America",
      "type": "TEAM_INVITE",
      "userIds": Array [
        3,
      ],
    },
  ],
}
`;
